src/server.c

```diff
@@ -6145,61 +6162,43 @@ void redisAsciiArt(void) {
     zfree(buf);
 }
 
-int changeBindAddr(void) {
-    /* Close old TCP and TLS servers */
-    closeSocketListeners(&server.ipfd);
-    closeSocketListeners(&server.tlsfd);
+int changeListener(const char *typename) {
+    int conn_index;
+    connListener *listener;
 
-    /* Bind to the new port */
-    if ((server.port != 0 && listenToPort(server.port, &server.ipfd) != C_OK) ||
-        (server.tls_port != 0 && listenToPort(server.tls_port, &server.tlsfd) != C_OK)) {
-        serverLog(LL_WARNING, "Failed to bind");
+    conn_index = connectionIndexByType(typename);
+    if (conn_index < 0)
+        serverPanic("Failed finding connection listener of %s", typename);
 
-        closeSocketListeners(&server.ipfd);
-        closeSocketListeners(&server.tlsfd);
-        return C_ERR;
-    }
-
-    /* Create TCP and TLS event handlers */
-    if (createSocketAcceptHandler(&server.ipfd, connAcceptHandler(CONN_TYPE_SOCKET)) != C_OK) {
-        serverPanic("Unrecoverable error creating TCP socket accept handler.");
-    }
-    if (createSocketAcceptHandler(&server.tlsfd, connAcceptHandler(CONN_TYPE_TLS)) != C_OK) {
-        serverPanic("Unrecoverable error creating TLS socket accept handler.");
-    }
-
-    if (server.set_proc_title) redisSetProcTitle(NULL);
-
-    return C_OK;
-}
-
-int changeListenPort(int port, socketFds *sfd, aeFileProc *accept_handler) {
-    socketFds new_sfd = {{0}};
+    listener = &server.listeners[conn_index];
+    listener->bindaddr = server.bindaddr;
+    listener->bindaddr_count = server.bindaddr_count;
+    listener->ct = connectionByType(typename);
+    if (!strcmp(typename, CONN_TYPE_SOCKET))
+        listener->port = server.port;
+    else if (!strcmp(typename, CONN_TYPE_TLS))
+        listener->port = server.tls_port;
 
     /* Close old servers */
-    closeSocketListeners(sfd);
+    closeSocketListeners(listener);
 
     /* Just close the server if port disabled */
-    if (port == 0) {
+    if (listener->port == 0) {
         if (server.set_proc_title) redisSetProcTitle(NULL);
         return C_OK;
     }
 
-    /* Bind to the new port */
-    if (listenToPort(port, &new_sfd) != C_OK) {
+    /* Re-create listener */
+    if (connListen(listener) != C_OK) {
```

Comment:
ok, i see (indeed no need to call closeSocketListeners on the same connListen that failed (listenToPort already does that when it fails).

but yes, i do think we need applyBind to clean after it in case it failed on a late stage (close the socket listener when the TLS listener failed).

IIRC, this mechanism was was set in place for the later rollback that's attempted.
i.e. if you change the port or bind at runtime and the listen fails, it attempts to roll back the actions, and re-bind to the old address / port.

p.s. i now notice that when createSocketAcceptHandler (in changeListener), in the past we used to fail gracefully, and now we panic. any reason for that?

Reply:
When createSocketAcceptHandler (in changeListener), we should fail gracefully instead of panic. I'll correct this.